src/experiment/malebolgia_spin_doctors.nim

```diff
@@ -0,0 +1,309 @@
+
+var
+  thr: array[ThreadPoolSize, Thread[int]]
+  chan: FixedChan
+  globalStopToken: Atomic[bool]
+  busyThreads: Atomic[int]
+  curretThread: Atomic[int]
+
+iterator threads(): int {.inline.} =
+  while true:
+    #var epoch = getMonoTime()
+    var n = curretThread.fetchAdd(1)
+    if n == FixedChanSize - 1:
+      curretThread.store(0, moRelaxed)
+    #echo inNanoseconds(getMonoTime() - epoch)
+    yield n
+
+proc send(item: sink PoolTask) =
+  # see deques.addLast:
+  var 
+    done = tsDone
+    empty = tsEmpty
+    ackchyually = 0
+
+  for i in threads():
+    # Test
+    if chan.board[i].load(moRelaxed) notin [done, empty]:
+      cpuRelax()
+      continue
+
+    # Test and Set
+    if chan.board[i].compareExchange(done,
+        tsFeeding, moAcquire, moRelease):
+      ackchyually = i
+      break
+
+    # Test and Set
+    if chan.board[i].compareExchange(empty,
+        tsFeeding, moAcquire, moRelease):
+      ackchyually = i
+      break
+    cpuRelax()
+
+  chan.data[ackchyually] = item
+  chan.board[ackchyually].store(tsToDo, moRelease)
+
+proc worker(i: int) {.thread.} =
+  var
+    item: PoolTask
+    todo = tsToDo
+    wip  = tsWip
+  while not globalStopToken.load(moRelaxed):
+    # Test
+    if chan.board[i].load(moRelaxed) != todo:
+      cpuRelax()
+      continue
+    # Test and Set
+    if not chan.board[i].compareExchange(todo,
+        wip, moRelaxed, moRelease):
+      cpuRelax()
+      continue
+
+    item = move chan.data[i]
+
+    if not item.m.stopToken.load(moRelaxed):
+      try:
+        atomicInc busyThreads
+        item.t.invoke(item.result)
+        atomicDec busyThreads
+        discard chan.board[i].compareExchange(wip,
+          tsDone, moRelease, moRelease)
+      except:
+        discard chan.board[i].compareExchange(wip,
+          tsFail, moRelease, moRelease)
+        withTTASLock item.m.L:
+          if item.m.error.len == 0:
+            let e = getCurrentException()
+            item.m.error = "SPAWN FAILURE: [" & $e.name & "] " & e.msg & "\n" & getStackTrace(e)
+
+    # but mark it as completed either way!
+    taskCompleted item.m[]
+
+proc setup() =
+  chan.L.store(false, moRelaxed)
+  globalStopToken.store(false, moRelaxed)
+  for i in 0..high(thr):
+    chan.board[i].store(tsEmpty, moRelaxed)
+    createThread[int](thr[i], worker, i)
+
+proc panicStop*() =
+  ## Stops all threads.
+  globalStopToken.store(true, moRelaxed)
+  joinThreads(thr)
+
+template spawnImplRes[T](master: var Master; fn: typed; res: T) =
+  if stillHaveTime(master):
+    if busyThreads.load(moRelaxed) < ThreadPoolSize:
+      taskCreated master
+      send PoolTask(m: addr(master), t: toTask(fn), result: addr res)
+    else:
+      res = fn
+
+template spawnImplNoRes(master: var Master; fn: typed) =
+  if stillHaveTime(master):
+    if busyThreads.load(moRelaxed) < ThreadPoolSize:
+      taskCreated master
+      send PoolTask(m: addr(master), t: toTask(fn), result: nil)
+    else:
+      fn
```

Comment:
I know isn't so simple, but:

If mainThread is used to perform some task, who is there to say to other threads what to do. 
We win one more thread and may lose -  they can be idle - seven.

```suggestion
    if isMainThread() or busyThreads.load(moRelaxed) < ThreadPoolSize:
      taskCreated master
      send PoolTask(m: addr(master), t: toTask(fn), result: nil)
    else:
      fn
```

Reply:
That's a good idea.